Angular 13 example for Relay

jemboh opened the pull request:
## Problem

As a web developer in 2022
I want to easily be able to use `js-waku` with Angular
So I can build awesome dapps with my favourite framework

## Solution

Here's an example send/receive messaging implementation with Angular, similar to the ReactJS one.

## Notes

- Resolves #320
- [Documentation PR](https://github.com/vacp2p/docs.wakuconnect.dev/pull/38)

D4nte commented on examples/relay-angular-chat/README.md:
```diff
@@ -0,0 +1,27 @@
+
+## Development server
+
+Run `ng serve` for a dev server. Navigate to `http://localhost:4200/`. The app will automatically reload if you change any of the source files.
```
For Markdown, best to use [semantic line breaks](https://sembr.org/). Makes editing easier, especially with git.

D4nte commented:
> - Resolves #320

We need docs.wakuconnect.dev to have the angular doc before we can close this issue #320.

D4nte commented on examples/relay-angular-chat/package.json:
```diff
@@ -0,0 +1,44 @@
+    "@angular/router": "~13.2.0",
+    "crypto-browserify": "^3.12.0",
+    "js-waku": "file:../../build/esm",
+    "protons": "^2.0.3",
```
We should move away from protons: https://github.com/status-im/js-waku/issues/335

I think it's ok to merge this PR with protons, it's more of an FYI.

D4nte commented on examples/relay-angular-chat/package.json:
```diff
@@ -0,0 +1,44 @@
+{
+  "name": "relay-angular-chat",
```
```suggestion
  "name": "@waku/relay-angular-chat",
```
FYI I control the @waku orga on npmjs so feel free to always prepend so we don't have bug hunters reserving the name on npmjs and claiming a bounty (I know you have set the `private: true`).

D4nte commented on examples/relay-angular-chat/src/app/app.component.ts:
```diff
@@ -0,0 +1,82 @@
+  // TODO: Probably better to create a Message interface and use that type here rather than any.
+  messages: any[] = [];
+  messageCount: number = 0;
+  // TODO: The other ignores are because of this. the Waku type isn't being picked up for some reason.
```
Maybe play around with the published package see if it helps? Doing this angular exercise is meant to dig up bugs like this one.

jemboh replied:
All of these have been resolved 🥳

D4nte commented on examples/relay-angular-chat/tsconfig.json:
```diff
@@ -0,0 +1,38 @@
+    "experimentalDecorators": true,
+    "moduleResolution": "node",
+    "importHelpers": true,
+    "target": "es2017",
```
I think the target may have to be es2020 due to the `BigInt` usage.

D4nte reviewed (commented):
Great job!

I see that the tax is not straightforward and you have unearth some issues around types import, which is exactly what we wanted to do with this issue: ensure that js-waku can work with Angular out-of-the-box.

jemboh commented:
> > * Resolves [Angular example  #320](https://github.com/status-im/js-waku/issues/320)
> 
> We need docs.wakuconnect.dev to have the angular doc before we can close this issue #320.

[PR raised](https://github.com/vacp2p/docs.wakuconnect.dev/pull/38)

D4nte commented:
Please add this folder to the [examples ci](https://github.com/status-im/js-waku/blob/master/.github/workflows/examples-ci.yml) to ensure we updated if we do a breaking change on js-waku API.

D4nte commented:
I am still getting an error. I added the folder to the examples CI so you can check from the CI whether all deps are properly committed

```
js-waku/examples/relay-angular-chat  angular-js-example-for-
relay ✔                                             6h36m  
▶ yarn start
yarn run v1.22.11
$ ng serve
✔ Browser application bundle generation complete.

Initial Chunk Files                                           | Names         |  Raw Size
vendor.js                                                     | vendor        |   5.89 MB | 
polyfills.js                                                  | polyfills     | 355.68 kB | 
styles.css, styles.js                                         | styles        | 173.29 kB | 
runtime.js                                                    | runtime       |  13.12 kB | 
main.js                                                       | main          |  12.73 kB | 

                                                              | Initial Total |   6.43 MB

Lazy Chunk Files                                              | Names         |  Raw Size
node_modules_js-waku_lib_waku_message_symmetric_node_js.js    | node          |   2.82 kB | 
node_modules_js-waku_lib_waku_message_symmetric_browser_js.js | browser       |   2.71 kB | 

Build at: 2022-02-16T03:48:19.906Z - Hash: ab5c4c00640f7e2f - Time: 17273ms

Error: ../../node_modules/libp2p-interfaces/dist/src/pubsub/index.d.ts:328:26 - error TS7016: Could not find a declaration file for module 'bl'. '/home/froyer/src/status-im/js-waku/node_modules/bl/bl.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/bl` if it exists or add a new declaration (.d.ts) file containing `declare module 'bl';`

328 type BufferList = import('bl');
                             ~~~~


Error: ../../node_modules/libp2p-interfaces/dist/src/stream-muxer/types.d.ts:1:24 - error TS7016: Could not find a declaration file for module 'bl/BufferList'. '/home/froyer/src/status-im/js-waku/node_modules/bl/BufferList.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/bl` if it exists or add a new declaration (.d.ts) file containing `declare module 'bl/BufferList';`

1 import BufferList from 'bl/BufferList'
                         ~~~~~~~~~~~~~~~


Error: ../../node_modules/libp2p-interfaces/dist/src/transport/types.d.ts:1:24 - error TS7016: Could not find a declaration file for module 'bl/BufferList'. '/home/froyer/src/status-im/js-waku/node_modules/bl/BufferList.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/bl` if it exists or add a new declaration (.d.ts) file containing `declare module 'bl/BufferList';`

1 import BufferList from 'bl/BufferList'
                         ~~~~~~~~~~~~~~~




** Angular Live Development Server is listening on localhost:4200, open your browser on http://localhost:4200/ **


✖ Failed to compile.

```

D4nte commented:
> I am still getting an error. I added the folder to the examples CI so you can check from the CI whether all deps are properly committed

I fixed the error by running `npm i --save-dev @types/bl` at the root of the repo.

D4nte reviewed (commented):
Looks good! 2 fix to do before merging:

- Check my questions re `ngInit`. I am not familiar with that but it's familar to React's `useEffect` then it happen every time the component is loaded which does not make it the best place to create the Waku node
- It seems that we could use a type dep at js-waku level (`bl`). It's strange. do you get the same issue when using the published js-waku package instead of local files?
- I added the folder to the example CI. It takes an eternity to run the test. It is using `npm` at the moment. I think we can switch it to `yarn` in a follow-up PR (or never if it does not impact the test) but we should try to keep CI execution short.

jemboh commented:
> I fixed the error by running `npm i --save-dev @types/bl` at the root of the repo.

Correct. I think I know why I missed this. I never deleted the `node_modules` from the root but I removed that dependency from the root `package.json`. It _is_ in the docs for a fresh angular project though. 

Added it back now.

jemboh commented on examples/relay-angular-chat/README.md:
```diff
@@ -0,0 +1,26 @@
+
+### Known issues
+
+There is a problem when using `npm` to install/run the Angular app.
```
Yes. I was going to ask why this was ignored in the root `.gitignore`. Removed it now.

D4nte commented:
>  > I fixed the error by running `npm i --save-dev @types/bl` at the root of the repo.
> 
> Correct. I think I know why I missed this. I never deleted the `node_modules` from the root but I removed that dependency from the root `package.json`. It _is_ in the docs for a fresh angular project though. 
> 
> Added it back now.
> 

I am still surprised we need to add it at the root. What is the behaviour when using js-waku from npmjs.com? Does installing the types in the angular project work?

jemboh commented on examples/relay-angular-chat/README.md:
```diff
@@ -0,0 +1,26 @@
+npm run build # Build js-waku
+cd examples/relay-reactjs-chat
+yarn   # Install dependencies for the web app
+yarn run start # Start development server to serve the web app on http://localhost:4200/
```
🩳

D4nte commented:
> > I am still surprised we need to add it at the root. What is the behaviour when using js-waku from npmjs.com? Does installing the types in the angular project work?
> 
> _I think_ because we're **not** using `js-waku` from `npmjs` but from `"file:../../build/esm"`, that's where it needs to be referenced from. 
> 
> When it's used from `npmjs` then it can be installed directly to the angular example folder. 
> 
> The `"file:../../build/esm"` was added for ci thought right?
> 

`file:` strategy is used because:
1. the library was not published when I built the first example
2. Enables easy testing/try out of changes in example
3. Forces us to keep examples and related doc up to date.

However, this is the 3rd time this strategy creates a problem. Ultimately, it may be better to simply use the npmjs.com package as it'll reproduce exactly how users are using the lib.

Also, regarding the reasons above in favour of `file:`:

1. Not valid anymore
2. Nothing stop a dev to change to `file:` locally for trying out a new change
3. Dependabot will ensure we don't forget to update the examples

So, moving forward, I propose you use the npmjs.com package in the angular example.
We can them move all examples to use published js-waku.

Finally, don't forget to add a dependabot entry for the new example.

jemboh commented on examples/relay-angular-chat/src/app/app.component.ts:
```diff
@@ -0,0 +1,81 @@
+  waku!: Waku;
+  wakuStatus: string = 'None';
+
+  ngOnInit(): void {
```
In this basic app, it's when the whole app starts as the 'app' is the component. 

In a 'real' angular app where there are multiple services and components to connect to waku, fetch and display the messages etc. The waku service would be a singleton and be injected (created) when needed in a component and destroyed when not, e.g. a chat window on the page. Or not. It would be up to the dApp developer.

D4nte replied:
ok thanks

jemboh commented on examples/relay-angular-chat/src/app/app.component.ts:
```diff
@@ -0,0 +1,81 @@
+  }
+  
+  ngOnDestroy(): void {
+    this.waku.relay.deleteObserver(this.processIncomingMessages, [this.contentTopic]);
```
> Is that matching ngOnInit?

Technically, yes.

> When is it called? 

So it's a bit of a cheat in that it's not called! It's more of an example of _how_ you would do it in Angular. If needed, I could split the app out in to multiple services and components and give examples of the different lifecycle methods?

> If the whole component is destroyed then there won't be a waku node anymore, would there be?

Correct but again, afaik, this component can't get destroyed as such in it's current state. Also, angular doesn't call the `ngOnDestroy` lifecycle method on page reload. You need to hook in to the DOM api and use `window.onbeforeunload` to call destroy. I've just pushed that up.

D4nte commented:
Remaining (just for the sake of clarity):
- use js-waku from npmjs.com instead of `file:`
- Fix examples-ci

jemboh commented:
Closing in favour of #549
